### PR TITLE
feat: Add support for X-Sentry-forwarded-for header

### PR DESCRIPTION
**Pull request by AbhiPrasad**
ref https://getsentry.atlassian.net/browse/OPS-3823
ref https://github.com/getsentry/relay/issues/2450

For users who use a [SDK tunnel](https://docs.sentry.io/platforms/javascript/troubleshooting/#using-the-tunnel-option) or proxy to route SDK events to Sentry, they often use the x-forwarded-for or [x-vercel-forwarded-for](https://vercel.com/docs/edge-network/headers#x-vercel-forwarded-for) to pass along the real client IP address to Sentry (so it can be stored with the Sentry event).

Current this doesn't work for SaaS Sentry because of restrictions we've placed on the infrastructure side.

~This PR adds support for reading from `X-Sentry-Forwarded-For` header. It is slightly worse that this is a proprietary header, but it unblocks this for our customers without having to suffer the same challenges x-forwarded-for does.~

With https://github.com/getsentry/ops/pull/8175 we're making changes to the infra to map the value in `X-Forwarded-For` to `X-Sentry-Forwarded-For` when the request gets forwarded from nginix to relay. This helps reduce the security concerns from blindly forwarding `X-Forwarded-For`, and preserves existing usage of `X-Forwarded-For`. In addition, self-hosted and local relay usage can still use `X-Forwarded-For` as usual.

### CODE REVIEW

**Review by jjbayer — APPROVED** [review]
> It is slightly worse that this is a proprietary header

Agree with this statement, but the code change LGTM.

**Comment by jan-auer** [comment]
https://getsentry.atlassian.net/browse/OPS-3823?focusedCommentId=89761

**Review by jan-auer — CHANGES_REQUESTED** [review]
The code change looks good. I would just like [this question](https://getsentry.atlassian.net/browse/OPS-3823?focusedCommentId=89761) addressed before moving ahead.

**Comment by AbhiPrasad** [comment]
Alright I think we're going to move forward as was outlined by Anton in the [JIRA ticket](https://getsentry.atlassian.net/browse/OPS-3823?focusedCommentId=89802), implemented here https://github.com/getsentry/ops/pull/8175.

@jan-auer sounds good to you?

**Review by jan-auer — APPROVED** [review]

File: relay-server/src/extractors/forwarded_for.rs [hunk 1]
```diff
@@ -11,12 +11,14 @@ pub struct ForwardedFor(String);
...
     const FORWARDED_HEADER: &str = "X-Forwarded-For";
     const VERCEL_FORWARDED_HEADER: &str = "X-Vercel-Forwarded-For";
+    const SENTRY_FORWARDED_HEADER: &str = "X-Sentry-Forwarded-For";
 
     /// We prefer the Vercel header because the normal one could get overwritten as explained here.
     /// `https://vercel.com/docs/concepts/edge-network/headers#x-vercel-forwarded-for`
     fn get_forwarded_for_ip(header_map: &HeaderMap) -> &str {
         header_map
             .get(Self::VERCEL_FORWARDED_HEADER)
+            .or_else(|| header_map.get(Self::SENTRY_FORWARDED_HEADER))
```
Comment (Dav1dde) [inline relay-server/src/extractors/forwarded_for.rs:28]: The question came up if we should value our own header over the vercel header, the conclusion was:

> 1. our infrastructure requires the "standard" x-forwarded-for to be trusted, so it puts the contents of just that header into the sentry version.
> 2. at the same time, x-vercel-forwarded-for is the more accurate header overall if it is there. so it should be preferred over the other two.

Can you add an explanation  to comment, so somebody in the future knows why the sentry header is not the highest priority (like there is already an explanation why the vercel header is preferred over the standard X-Forwarded-For header)?

Comment (AbhiPrasad) [reply relay-server/src/extractors/forwarded_for.rs:28]: Done with 3a97475d1e6d11ad6724c935a39e62fa61489582

**Pull request merged**